src/libgt2/libgt2.cpp

```diff
@@ -0,0 +1,268 @@
+    progress_data* pd = (progress_data*)payload;
+    pd->completed_steps = cur;
+    pd->total_steps = tot;
+    pd->path = path;
+    print_progress(pd);
+}
+
+// Callback for clone authentication - will be used when password is not set in repo_url
+// Does not work with LFS download as it requires additional authentication when password is not set in repository url
+int cred_acquire_cb(git_credential** out,
+    const char* url,
+    const char* username_from_url,
+    unsigned int allowed_types,
+    void* payload) {
+    char *username = NULL, *password = NULL;
+    int error = -1;
+
+    fprintf(stdout, "Authentication is required for repository clone.\n");
+    if (allowed_types & GIT_CREDENTIAL_USERPASS_PLAINTEXT) {
+        const char* env_cred = std::getenv("HF_TOKEN");
+#ifdef __linux__
+        if (env_cred) {
+            username = strdup(env_cred);
+            password = strdup(username);
+        } else {
+            fprintf(stderr, "HF_TOKEN env variable is not set.\n");
+            return -1;
+        }
+#elif _WIN32
+        if (env_cred) {
+            username = _strdup(env_cred);
+            password = _strdup(username);
+        } else {
+            fprintf(stderr, "HF_TOKEN env variable is not set.\n");
+            return -1;
+        }
+#endif
+        error = git_credential_userpass_plaintext_new(out, username, password);
+        if (error < 0) {
+            fprintf(stderr, "Creating credentials failed.\n");
+            error = -1;
+        }
+    } else {
+        fprintf(stderr, "Only USERPASS_PLAINTEXT supported in OVMS.\n");
+        return 1;
+    }
+
+    free(username);
+    free(password);
+    return error;
+}
+
+// C-style callback functions section used in libgt2 library ENDS ********************************
+bool HfDownloader::isPullHfModelModeOn() {
+    return this->pullHfModelMode;
+}
+
+bool HfDownloader::CheckIfProxySet() {
+    const char* envCred = std::getenv("https_proxy");
+    if (envCred)
+        return true;
+    return false;
+}
+
+bool HfDownloader::CheckIfTokenSet() {
+    const char* envCred = std::getenv("HF_TOKEN");
+    if (envCred)
+        return true;
+    return false;
+}
+
+std::string HfDownloader::GetRepositoryUrlWithPassword(std::string& hfEndpoint) {
+    std::string passRepoUrl = "https://";
+
+    const char* envCred = std::getenv("HF_TOKEN");
+    if (envCred) {
+        std::string cred = std::string(envCred);
+        passRepoUrl += cred + ":" + cred + "@";
+    } else {
+        SPDLOG_DEBUG("HF_TOKEN environment variable not set");
+    }
+
+    passRepoUrl += hfEndpoint + this->sourceModel;
+
+    return passRepoUrl;
+}
+
+std::string HfDownloader::GetHfEndpoint() {
+    const char* envCred = std::getenv("HF_ENDPOINT");
+    std::string hfEndpoint = "huggingface.co";
+    if (envCred) {
+        hfEndpoint = std::string(envCred);
+    } else {
+        SPDLOG_DEBUG("HF_ENDPOINT environment variable not set");
```

Comment:
Is this reason to stop OVMS if we require to pull but endpoint is not set?

Reply:
We use default endpoint huggingface.co